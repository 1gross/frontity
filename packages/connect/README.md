## Description
The state manager of Frontity. Similar to MobX but much more lightweight because it's based on ES6 proxies.

You can find more information about it at [Frontity docs](https://docs.frontity.org/javascript-and-and-react/react-basic#state-manager-frontity-connect).
