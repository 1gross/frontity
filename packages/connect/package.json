{
  "name": "@frontity/connect",
  "version": "0.1.0",
  "description": "A state manager that's easy to use and hackable at the same time.",
  "keywords": [
    "frontity",
    "connect"
  ],
  "homepage": "https://frontity.org",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/frontity/frontity.git"
  },
  "bugs": {
    "url": "https://community.frontity.org"
  },
  "scripts": {
    "test:ci": "jest",
    "test": "jest --watch",
    "test:inspect": "node --inspect ./node_modules/.bin/jest --watch --no-cache --runInBand"
  },
  "dependencies": {
    "recursive-iterator": "^3.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.4.4",
    "@babel/preset-env": "^7.4.4",
    "@types/jest": "^24.0.11",
    "babel-jest": "^24.8.0",
    "jest": "^24.7.1",
    "ts-jest": "^24.0.2",
    "typescript": "^3.4.2"
  }
}
