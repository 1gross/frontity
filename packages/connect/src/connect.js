import { createContext, useContext, createElement } from "react";
import { view } from "@risingstack/react-easy-state";

/**
 * A React context that stores the store of Frontity Connect.
 */
const context = createContext();

/**
 * The Provider of the store context, to be used in the root of the React
 * application.
 */
export const Provider = context.Provider;

/**
 * Connects a React component with the Frontity connect store.
 *
 * @param Component The React component to be connected
 * @returns The same Component, but with the store included in the props.
 */
export const connect = (Component) => {
  const isStatelessComp = !(
    Component.prototype && Component.prototype.isReactComponent
  );

  if (isStatelessComp) {
    return view((props) => {
      const store = useContext(context);
      return Component({ ...props, ...store });
    });
  } else {
    class ConnectedComponent extends Component {
      static contextType = context;

      render() {
        return createElement(Component, {
          ...this.props,
          ...this.context,
        });
      }
    }

    return view(ConnectedComponent);
  }
};
