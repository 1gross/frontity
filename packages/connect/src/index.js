export { observe, unobserve } from "./observer";
export { observable, isObservable, raw, clone } from "./observable";
export { createStore, getSnapshot } from "./create-store";
export { connect as default, Provider, useConnect } from "./connect";
export { batch } from "./scheduler";
