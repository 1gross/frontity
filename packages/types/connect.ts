import Package from "./package";
import { ResolveState, Omit } from "./utils";

// Resolves the actions for things like Action and Derived.
export type ResolveActions<Actions extends Package["state"]> = {
  [P in keyof Actions]: Actions[P] extends (
    ...a: any
  ) => (arg: infer Arg) => void // Turns "state => args => {}" into "args => {}"
    ? (arg: Arg) => void // Turns "state => {}" into "() => {}"
    : Actions[P] extends (state: Package["state"]) => any
    ? () => void
    : ResolveActions<Actions[P]>
};

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export default Connect;
