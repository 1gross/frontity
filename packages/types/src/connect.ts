import Package from "./package";
import { ResolveState, Omit, ResolveActions } from "./utils";

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export type ExternalProps<T extends Package> = Omit<
  T,
  "state" | "actions" | "libraries" | "roots" | "fills"
>;

export default Connect;
