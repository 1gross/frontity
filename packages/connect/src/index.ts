import { Store, State, ProxifyOptions, Context, Actions } from "../types";

export const proxifyState = <StoreType extends Store>(
  rawStore: StoreType,
  context?: Context,
  options?: ProxifyOptions
): State<StoreType> => {
  return rawStore.state as State<StoreType>; // Overwrite type until it's implemented.
};

export const proxifyActions = <StoreType extends Store>(
  rawStore: StoreType,
  context?: Context,
  options?: ProxifyOptions
): Actions<StoreType> => {
  return rawStore.actions as Actions<StoreType>; // Overwrite type until it's implemented.
};
