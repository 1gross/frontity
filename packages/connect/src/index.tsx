/* eslint-disable @typescript-eslint/no-unused-vars */
import React, { ComponentType, FunctionComponent } from "react";
import { Store, State, ProxifyOptions, Actions, FilterStore } from "../types";

export const proxifyState = <StoreType extends Store>(
  rawStore: StoreType,
  options?: ProxifyOptions
): State<StoreType> => {
  return rawStore.state as State<StoreType>; // Overwrite type until it's implemented.
};

export const proxifyActions = <StoreType extends Store>(
  rawStore: StoreType,
  options?: ProxifyOptions
): Actions<StoreType> => {
  return rawStore.actions as Actions<StoreType>; // Overwrite type until it's implemented.
};

const connect = <Props extends Store>(
  Component: ComponentType<Props>,
  options?: ProxifyOptions
): FunctionComponent<FilterStore<Props>> => {
  return Component as FunctionComponent<FilterStore<Props>>;
};

export const Provider: React.FC<{ value: Store }> = ({ children }) => (
  <>{children}</>
);

export default connect;
