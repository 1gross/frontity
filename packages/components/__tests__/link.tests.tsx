import React from "react";
import { create } from "react-test-renderer";
import { createStore, Provider } from "../../connect";
import Link from "../link";

let store;

beforeEach(() => {
  store = createStore({
    state: {},
    actions: {
      router: {
        set(link) {
          return link;
        },
      },
    },
  });
});

describe("Link", () => {
  test("should render a regular link", () => {
    const LinkComponent = create(
      <Provider value={store}>
        <Link link="/">This is a link</Link>
      </Provider>
    );

    expect(LinkComponent.toJSON()).toMatchSnapshot();
  });

  test("should add classname if provided", () => {
    const LinkComponent = create(
      <Provider value={store}>
        <Link link="/" className="custom-classname">
          This is a link
        </Link>
      </Provider>
    );

    expect(LinkComponent.toJSON()).toMatchSnapshot();
  });

  test("calls action.router.set", () => {
    const LinkComponent = create(
      <Provider value={store}>
        <Link link="/">This is a link</Link>
      </Provider>
    );
  });
});
