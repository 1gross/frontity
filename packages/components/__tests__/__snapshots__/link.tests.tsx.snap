// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should add classname if provided 1`] = `
<a
  className="custom-classname"
  href="/"
  onClick={[Function]}
  target="_self"
>
  This is a link
</a>
`;

exports[`Link should render a regular link 1`] = `
<a
  className=""
  href="/"
  onClick={[Function]}
  target="_self"
>
  This is a link
</a>
`;
